src/mtcp/mtcp_restart.c

```diff
@@ -858,42 +859,54 @@ static int read_one_memory_area(int fd)
       if (area.prot & PROT_WRITE) flags = O_RDWR;
     }
 
-    if (area.prot & MAP_SHARED) {
-      imagefd = mtcp_sys_open (area.name, flags, 0);  // Can we open file?
-      if (imagefd < 0 && mtcp_sys_errno == ENOENT) {
-        // File doesn't exist.  Do we have perm to create it and write data?
-        imagefd = mtcp_sys_open (area.name, O_CREAT|O_RDWR, 0);
-        if (imagefd >= 0) {
-          mtcp_sys_unlink(area.name);
-        } else {
-          // We don't have permission to re-create shared file.
-          // Open it as anonymous private, and hope for the best.
-          // FIXME: Or maybe we do have permission to re-create shared file,
-          // but it requires us to also re-create the parent directory.
-          area.flags ^= MAP_SHARED;
-          area.flags |= MAP_PRIVATE;
-          area.flags |= MAP_ANONYMOUS;
-          if (area.prot & PROT_WRITE) {
-            area.prot ^= PROT_WRITE;
-            MTCP_PRINTF("Found old underlying file %s\n"
-                        " that no longer exists, but it's mapped shared,"
-                        " with write permission,\n and we don't have permission"
-                        " to re-create it.  We will map it read-only\n"
-                        " and hope for the best.\n", area.name);
-          }
-          goto read_data;
-        }
-      }
+    int can_create_area = 1;
```

Comment:
For the MAP_PRIVATE case, we don't need to create the directory tree. If a file is missing, it will automatically be updated to MAP_ANONYMOUS and restart will proceed as expected. Recreating the directory tree in this case doesn't help. In fact, it's less optimal because we are unnecessarily creating the directory tree, recreating the file and populating it's contents from the file. However, once the file has been created, it won't be used again by the restarted process.